project/src/main/kotlin/org/ionproject/core/accessControl/AccessServices.kt

```diff
@@ -24,18 +34,75 @@ class AccessServices(private val authRepo: AuthRepo, private val tokenGenerator:
     }
 
     /**
-     * A possible optimization is to insert the hash in the request
-     * at the moment its first checked during the PDP
+     * Revokes the presented token reference in the body
+     */
+    fun revokeToken(tokenHash: String) {
+        authRepo.revokeToken(tokenHash)
+    }
+
+    /**
+     * Revokes the derived tokens of the presented token reference in the body
+     * e.g. revoke all import url's originated by the read token presented
+     */
+    fun revokeChild(tokenHash: String) {
+        authRepo.revokeChild(tokenHash)
+    }
+
+    /**
+     * Revokes the derived tokens and the presented token
+     */
+    fun revokePresentedAndChild(tokenHash: String) {
+        authRepo.revokePresentedAndChild(tokenHash)
+    }
+
+    /**
+     * Checks if class calendar exists, if it does generate or share token
+     */
+    fun generateImportClassCalendar(
+        cid: Int,
+        calterm: String,
+        query: MultiValueMap<String, String>,
+        parameterPath: String,
+        tokenHash: String
+    ) : String {
+        val calendar = calRepo.getClassCalendar(cid, calterm, query)
+
+        if(calendar != null)
+            return generateImportToken(parameterPath, tokenHash)
+        else
+            throw BadRequestException("The class calendar you tried to import doesn't exist.")
+    }
+
+    /**
+     * Checks if class section calendar exists, if it does generate or share token
      */
-    fun revokeToken(token: String) {
-        val hash = tokenGenerator.getHash(tokenGenerator.decodeBase64url(token))
-        authRepo.revokeToken(hash)
+    fun generateImportClassSectionCalendar(
+        sid: String,
+        calterm: String,
+        cid: Int,
+        query: MultiValueMap<String, String>,
+        parameterPath: String,
+        tokenHash: String
+    ) : String {
+
+        val calendar = calRepo.getClassSectionCalendar(
```

Comment:
If such check wasn't made an malicious user could issue millions of different import URL's to "Not Found" resources...